initrd/bin/gui-init

```diff
@@ -92,17 +129,243 @@ update_checksums()
+
+        whiptail --title 'DISK UNLOCK KEY ADVICE 2/3' --msgbox \
+          "Otherwise, typing your Disk Recovery Key passphrase at boot puts you\nat risk. Someone could clone your disk and type that witnessed\npassphrase and access your data both locally and from a cloned\ndisk.\n\nCoupled with a strong passphrase, this improves your data privacy.\nCONSIDER YOURSELF WARNED.\n\nPlease hit Enter to continue" 30 60
+
+        whiptail --title 'DISK UNLOCK KEY ADVICE 3/3' --msgbox \
+          "To set Disk Unlock Key, released by the TPM only if firmware integrity\n is attested:\n\nSelect from Main menu:\n\nAdvanced Options menu -> Other Boot Options -> Show OS boot menu\n1-Then select the first boot option proposed (dynamic option)\n2-Then,answer Y when asked to add Disk Unlock Key to the TPM.\n\nYou will be prompted to set one in the following steps.\n\nPlease hit Enter to continue" 30 60
+        return 
+      else
+        #Deal with corner case where after inserting key in previous step, HOTP seal fails with provided good password.
+        #In that case, we regenerate HOTP and TOTP
+        totp_confirm="g"
+        echo $totp_confirm > /tmp/whiptail
+      fi
+    fi
+  else
+    reset_disk_unlock_key
+    return
+  fi  
+}
+reset_disk_unlock_key()
+{
+  mount_boot
+  
+  if [ "$CONFIG_TPM" = n ]; then
+    TOTP="NO TPM"
+  else
+    TOTP=`unseal-totp`
+    if [ $? -ne 0 ]; then
+      echo "Error generating TOTP code. We will regenerate it now."
+      update_totp
+      continue
+    elif [ -x /bin/libremkey_hotp_verification ]; then
+      HOTP=`unseal-hotp`
+      enable_usb
+      if ! libremkey_hotp_verification info ; then
+        whiptail $CONFIG_WARNING_BG_COLOR --clear --title 'WARNING: Please Insert Your LibremKey/Nitrokey' --msgbox "Your LibremKey/Nitrokey was not detected.\n\nPlease insert your LibremKey/Nitrokey" 30 90
+      fi
+      # Don't output HOTP codes to screen, so as to make replay attacks harder
+      libremkey_hotp_verification check "$HOTP"
+      case "$?" in
+        0 )
+          HOTP="Success"
+        ;;
+        4 )
+          HOTP="Invalid code"
+          MAIN_MENU_BG_COLOR=$CONFIG_ERROR_BG_COLOR
+        ;;
+        * )
+          HOTP="Error checking code, Insert LibremKey/Nitrokey and retry"
+          MAIN_MENU_BG_COLOR=$CONFIG_WARNING_BG_COLOR
+        ;;
+      esac
+    else
+      HOTP='N/A'
+    fi
+      
+    if [ $HOTP = "Success" ];then
+      mount -o remount,rw /boot
+      rm /boot/reset_disk_unlock_key
+      mount -o remount,ro /boot
+      verify_global_hashes
+      kexec-select-boot -m -b /boot -c "grub.cfg" -g
+      continue
+    else
+      echo -e "HOTP status: $HOTP\n\nHit Enter to regenerate new TOTP/HOTP codes."
+      read
+      update_totp
+    fi
+    continue
+  fi
+}
+
+mount_boot
+if [ -e /boot/oem ];then
```

Comment:
@kylerankin : good enough if it's in /etc/functions? I still think all functions should be centralized.

Reply:
It *probably* is good enough if it's there as long as it's well-tested. That's one reason why I initially held back from putting the GUI file selection function there (but I think it's ready to go there now that we call it from a number of scripts). 

We just don't want a syntax error to mean that every script that sources that file breaks.

Reply:
@kylerankin The file selection is called from gui-init only anyway?

Reply:
It's now called from gui-init but it is also used both in gpg-gui and flash-gui.

Reply:
@kylerankin : I have let the file_selector code to be locally defined under initrd/bin/config-gui.sh and initrd/bin/gpg-gui.sh but you're right, those could be moved to /etc/functions since they are exact duplicate of each other. Will do.

Reply:
@kylerankin : removed in both scripts